module/VuFind/src/VuFind/Db/Table/LoginToken.php

```diff
@@ -0,0 +1,192 @@
+     * @param int    $expires   Token lifetime in days
+     * @param string $sessionId Session associated with the token
+     *
+     * @return LoginToken
+     */
+    public function saveToken(
+        int $userId,
+        string $token,
+        string $series,
+        string $browser = '',
+        string $platform = '',
+        int $expires = 0,
+        string $sessionId = ''
+    ) {
+        $row = $this->createRow();
+        $row->token = hash('sha256', $token);
+        $row->series = $series;
+        $row->user_id = $userId;
+        $row->last_login = date('Y-m-d H:i:s');
+        $row->browser = $browser;
+        $row->platform = $platform;
+        $row->expires = $expires;
+        $row->last_session_id = $sessionId;
+        $row->save();
+        return $row;
+    }
+
+    /**
+     * Check if a login token matches one in database.
+     *
+     * @param array $token array containing user id, token and series
+     *
+     * @return mixed
+     * @throws LoginTokenException
+     */
+    public function matchToken(array $token)
+    {
+        $row = $this->getBySeries($token['series'], $token['user_id']);
+        if ($row && hash_equals($row['token'], hash('sha256', $token['token']))) {
+            if (time() > $row['expires']) {
+                $row->delete();
+                return false;
+            }
+            return $row;
+        } elseif ($row) {
+            // Matching series and user id found, but token does not match
+            // throw exception
+            throw new LoginTokenException('Token does not match');
+        }
+        return false;
+    }
+
+    /**
+     * Delete all tokens in a given series
+     *
+     * @param string $series series
+     * @param int    $userId User identifier
+     *
+     * @return void
+     */
+    public function deleteBySeries(string $series, int $userId)
+    {
+        $this->delete(['user_id' => $userId, 'series' => $series]);
+    }
+
+    /**
+     * Delete all tokens for a user
+     *
+     * @param int $userId user identifier
+     *
+     * @return void
+     */
+    public function deleteByUserId(int $userId)
+    {
+        $this->delete(['user_id' => $userId]);
+    }
+
+    /**
+     * Get tokens for a given user
+     *
+     * @param int $userId User identifier
+     *
+     * @return \VuFind\Db\Row\LoginToken
+     */
+    public function getByUserId(int $userId)
+    {
+        $callback = function ($select) use ($userId) {
+            $select->where->equalTo('user_id', $userId);
+            $select->order('last_login DESC');
+        };
+        return $this->select($callback);
+    }
+
+    /**
+     * Get token by series
+     *
+     * @param string $series Series identifier
+     * @param int    $userId User identifier
+     *
+     * @return LoginToken
+     */
+    public function getBySeries(string $series, int $userId)
+    {
+        return $this->select(['user_id' => $userId, 'series' => $series])->current();
```

Comment:
Does it matter that this result set is not sorted? Is there a reason to sort by last_login DESC when using user_id, but not when using series?

Reply:
When getting tokens for display, getByUserId is used, so that is why it is sorted. It should not matter that getBySeries is not sorted